libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt

```diff
@@ -27,63 +25,42 @@ internal class TimeBasedNextVoiceInstructionsProvider(
             }
 
         val voiceInstructions = mutableListOf<VoiceInstructions>()
-        fillCurrentStepVoiceInstructions(
-            currentStep,
-            progress.stepDistanceRemaining,
-            voiceInstructions
-        )
         var cumulatedTime = progress.stepDurationRemaining
+        val currentStepInstructions = currentStep.voiceInstructions()?.filter { instruction ->
+            val distanceAlongGeometry = instruction.distanceAlongGeometry()
+            distanceAlongGeometry != null &&
+                distanceAlongGeometry <= progress.stepDistanceRemaining
+        }
+        if (currentStepInstructions != null) {
+            voiceInstructions.addAll(currentStepInstructions)
+        }
 
-        // fill next steps
         var currentStepIndex = progress.stepIndex
         var currentLegIndex = progress.legIndex
         while (cumulatedTime < observableTimeSeconds) {
-            if (isLastStep(currentStepIndex, legSteps)) {
+            if (currentStepIndex + 1 < (legSteps?.size ?: 0)) {
+                currentStep = legSteps!![currentStepIndex + 1]
+                currentStepIndex++
+            } else {
                 currentStepIndex = 0
-                if (isLastLeg(currentLegIndex, legs)) {
-                    break
-                } else {
+                if (currentLegIndex + 1 < legs.size) {
                     legSteps = legs[currentLegIndex + 1].steps()
                     currentLegIndex++
                     if (legSteps.isNullOrEmpty()) {
                         continue
                     } else {
                         currentStep = legSteps.first()
                     }
+                } else {
+                    break
                 }
-            } else {
-                currentStep = legSteps!![currentStepIndex + 1]
-                currentStepIndex++
             }
             currentStep.voiceInstructions()?.let { voiceInstructions.addAll(it) }
             cumulatedTime += currentStep.duration()
         }
         return voiceInstructions
     }
 
-    private fun fillCurrentStepVoiceInstructions(
```

Comment:
Looks like you only inlined the functions in this class and inverted `if` statements, was it really necessary? This makes it harder to review.